packages/astro/src/core/app/index.ts

```diff
@@ -250,12 +253,53 @@ export class App {
 		return pathname;
 	}
 
+	#redirectTrailingSlash(pathname: string): string | undefined {
+		const { trailingSlash } = this.#manifest;
+
+		// Ignore root and internal paths
+		if (pathname === '/' || pathname.startsWith('/_')) {
+			return;
+		}
+
+		// Redirect multiple trailing slashes to collapsed path
+		if (MANY_SLASHES.test(pathname)) {
+			const path = pathname.replace(MANY_SLASHES, trailingSlash === 'never' ? '' : '/');
+			return path ? path : '/';
+		}
```

Comment:
This function should be moved inside `internal-helpers` (the `MANY_SLASHES` constant, too) since it's been used in multiple places